mage_integrations/mage_integrations/sources/dynamodb/__init__.py

```diff
@@ -0,0 +1,182 @@
+            column_name = column_schema.get(ATTRIBUTE_NAME)
+            column_type = column_schema.get(ATTRIBUTE_TYPE)
+            column_types = []
+            if column_type == 'S':
+                column_types.append(COLUMN_TYPE_STRING)
+            elif column_type == 'N':
+                column_types.append(COLUMN_TYPE_NUMBER)
+            elif column_type == 'B':
+                column_types.append(COLUMN_TYPE_OBJECT)
+            else:
+                self.logger.error(f'Unknown column type {column_type} for column {column_name}')
+
+            properties[column_name] = dict(
+                    type=column_types,
+                )
+        schema = Schema.from_dict(dict(
+                properties=properties,
+                type='object',
+            ))
+        metadata = get_standard_metadata(
+                key_properties=key_props,
+                schema=schema.to_dict(),
+                stream_id=table_name,
+            )
+        catalog_entry = CatalogEntry(
+                key_properties=key_props,
+                metadata=metadata,
+                schema=schema,
+                stream=table_name,
+                tap_stream_id=table_name,
+            )
+        return catalog_entry
+
+    def discover(self, streams: List[str] = None) -> Catalog:
+        '''
+        Read streams in DynamoDB.
+        '''
+        client = self.build_client()
+        outputs = []
+        if streams:
+            # Check if streams available
+            for stream in streams:
+                catalog_entry = self.discover_stream(client, stream)
+                if catalog_entry is not None:
+                    outputs.append(catalog_entry)
+        return Catalog(outputs)
+
+    def build_client(self):
+        '''
+        Build DynamoDB client.
+        '''
+        config = Config(
+           retries={
+              'max_attempts': 10,
+              'mode': 'standard',
+           },
+        )
+
+        return boto3.client(
+            'dynamodb',
+            aws_access_key_id=self.config[AWS_ACCESS_KEY_ID],
+            aws_secret_access_key=self.config[AWS_SECRET_ACCESS_KEY],
+            config=config,
+            region_name=self.region,
+        )
+
+    def test_connection(self) -> None:
+        client = self.build_client()
+        client.describe_endpoints()
+
+    def scan_table(self, table_name):
+        '''
+        Scan records in one table.
+        '''
+        scan_params = {
+            'TableName': table_name,
+            'Limit': 1000
+        }
+
+        client = self.build_client()
+        has_more = True
+
+        while has_more:
+            result = client.scan(**scan_params)
+            yield result
+
+            if result.get(LAST_EVALUATED_KEY):
+                scan_params[EXCLUSIVE_STAR_KEY] = result[LAST_EVALUATED_KEY]
+
+            has_more = result.get(LAST_EVALUATED_KEY, False)
+
+    def load_data(
+        self,
+        stream,
+        bookmarks: Dict = None,
```

Comment:
we should probably take the bookmarks into account. you can look at other sources to see how bookmarks work

Reply:
Added the logic with bookmark.

So if `bookmarks` contains the `LAST_EVALUATED_KEY`, we will scan the table beginning with `LAST_EVALUATED_KEY`. Meanwhile, anytime we get result from scan_table, we record the LAST_EVALUATED_KEY as the beginning for next round.

Reply:
@matrixstone the bookmarks here will look something like based on the `state` of the source.
```
bookmarks: {
  bookmark_property_1: xxx,
  bookmark_property_2: xxx,
}
```

also, users can select what their bookmark properties are, so we'll want to filter by that as well.

overall, the bookmarking logic makes sense, but we just need to fix how we read from the bookmarks parameter, and add some logic for the bookmark properties. here's the sql one for reference: https://github.com/mage-ai/mage-ai/blob/master/mage_integrations/mage_integrations/sources/sql/base.py#L328

Reply:
Updated the PR to include two functions `_get_bookmark_properties_for_stream` and `__get_bookmarks_for_stream` in `dynamodb/__init__.py`.  Also when reading from bookmark in `load_data`, we will check if it is in `bookmark_properties`

Reply:
Updated the bookmark logic.

Once I set the `valid_replication_keys` with `key properties`, the `bookmarks` dict contains those keys and its values. At the same time, the format of bookmarks is same as the `last evaluated key` from dynamodb, so we can use it to query incremental data directly.

I tested it with Mage UI. Attached the screenshot in the description in this PR.